scripts/build.js

```diff
@@ -5,28 +5,132 @@
 const os = require('os');
 const process = require("process");
 const cmake = require("cmake-js");
+const axios = require("axios");
+const path = require("path");
+const tar = require('tar');
+const fs = require("fs-extra");
+const { v4: uuidv4 } = require('uuid');
+const checksum = require('checksum');
 
-let options = {
-    CMAKE_EXPORT_COMPILE_COMMANDS: true,
-    CMAKE_JS_PLATFORM: os.platform,
-    CMAKE_JS_ARCH: os.arch,
-    BUILD_TESTING: 'OFF',
-    CMAKE_INSTALL_PREFIX: 'crt/install',
-    CMAKE_PREFIX_PATH: 'crt/install',
+async function download_file(fileUrl, outputLocationPath) {
+    const writer = fs.createWriteStream(outputLocationPath);
+    return axios({
+        method: 'get',
+        url: fileUrl,
+        responseType: 'stream',
+    }).then(response => {
+        return new Promise((resolve, reject) => {
+            response.data.pipe(writer);
+            let error = null;
+            writer.on('error', err => {
+                error = err;
+                writer.close();
+                reject(err);
+            });
+            writer.on('close', () => {
+                if (!error) {
+                    resolve(true);
+                }
+            });
+        });
+    });
 }
 
-// Convert any -D arguments to this script to cmake -D arguments
-for (const arg of process.argv) {
-    if (arg.startsWith('-D')) {
-        const option = arg.substring(2).split('=')
-        options[option[0]] = option[1]
+async function check_checksum(url, loacl_file) {
+    return axios({
+        method: 'get',
+        url: url,
+        responseType: 'text',
+    }).then(response => {
+        return new Promise((resolve, reject) => {
+            checksum.file(loacl_file, function (err, sum) {
+                if (err) {
+                    reject(err);
+                }
+                if (sum === response.data.slice(0, -1)) {
+                    resolve()
+                }
+                else {
+                    reject(new Error("source code checksum mismatch"))
+                }
+            })
+        });
+    })
+}
+
+async function fetch_native_code(url, version, path) {
+    const source_URL = url + "aws-crt-" + version + "-source.tgz"
+    const tarball_path = path + "source.tgz"
+    return new Promise((resolve, reject) => {
+        download_file(source_URL, tarball_path).then(() => {
+            // download checksum
+            const source_checksum_URL = url + "aws-crt-" + version + "-source.sha1"
+            check_checksum(source_checksum_URL, tarball_path)
+            // check_checksum(source_checksum_URL, "/Users/dengket/Downloads/aws-crt-1.9.2-source.tgz")
```

Comment:
something is wrong! The cloudfront link works perfectly on my browser, downloaded the latest .tgz file. However, the nodejs axios gives me a cached or old version of the file? I checked the returned header, it apparently not the latest one. I don't really know what happens.

Reply:
I just tested today, it works now🙃

Reply:
I think we can ignore the caching problem, as when we publish for real, we will have different version number every time. It will be different urls, and get around the caching issue.